src/accountsdb/buffer_pool.zig

```diff
@@ -0,0 +1,1074 @@
+    total_requested_bytes -= @sizeOf(usize) * 3; // hashmap header
+
+    try std.testing.expect(total_requested_bytes % frame_count == 0);
+
+    // metadata should be small!
+    // As of writing, all metadata (excluding eviction_lfu, including frame_map)
+    // is 50 bytes or ~9% of memory usage at a frame size of 512, or 50MB for a
+    // million frames.
+    try std.testing.expect((total_requested_bytes / frame_count) - 512 <= 64);
+}
+
+test "BufferPool filesize > frame_size * num_frames" {
+    const allocator = std.testing.allocator;
+
+    const file = try std.fs.cwd().openFile("data/test-data/test_account_file", .{});
+    defer file.close();
+    const file_id = FileId.fromInt(1);
+
+    const num_frames = 200;
+
+    const file_size = (try file.stat()).size;
+    if (file_size < FRAME_SIZE * num_frames) @panic("file too small for valid test");
+
+    var bp = try BufferPool.init(std.heap.page_allocator, num_frames);
+    defer bp.deinit(std.heap.page_allocator);
+
+    // can't read buffer larger than total size of the buffer pool
+    const read_whole = bp.read(allocator, file, file_id, 0, @intCast(file_size - 1));
+    try std.testing.expectEqual(error.InvalidArgument, read_whole);
+
+    // file_size > total buffers size => we evict as we go
+    var offset: u32 = 0;
+    while (offset < file_size) : (offset += FRAME_SIZE) {
+        // when we've already filled every frame, we evict as we go
+        // => free list should be empty
+        if (offset >= FRAME_SIZE * num_frames) {
+            try std.testing.expectEqual(0, bp.free_list.len.raw);
+        } else {
+            try std.testing.expect(bp.free_list.len.raw > 0);
+        }
+
+        const read_frame = try bp.read(
+            allocator,
+            file,
+            file_id,
+            offset,
+            offset + FRAME_SIZE,
+        );
+
+        try std.testing.expectEqual(1, read_frame.indices.len);
+
+        const frame: []const u8 = bp.frames[
+            read_frame.indices[0]
+        ][0..bp.frames_metadata.size[
+            read_frame.indices[0]
+        ]];
+
+        var frame2: [FRAME_SIZE]u8 = undefined;
+        const bytes_read = try file.preadAll(&frame2, offset);
+        try std.testing.expectEqualSlices(u8, frame2[0..bytes_read], frame);
+        read_frame.deinit(allocator);
+    }
+}
+
+test "BufferPool random read" {
+    const allocator = std.testing.allocator;
+
+    const file = try std.fs.cwd().openFile("data/test-data/test_account_file", .{});
+    defer file.close();
+    const file_id = FileId.fromInt(1);
+
+    const num_frames = 200;
+
+    var bp = try BufferPool.init(allocator, num_frames);
+    defer bp.deinit(allocator);
+
+    const file_size: u32 = @intCast((try file.stat()).size);
+
+    var prng = std.Random.DefaultPrng.init(5083);
+
+    var reads: usize = 0;
+    while (reads < 5000) : (reads += 1) {
+        var gpa = std.heap.GeneralPurposeAllocator(.{}){};
```

Comment:
I wanted to use the gpa to maintain that every individual read (and corresponding deinit) never leaks - wanted to avoid potential bug where leaks happen between them. Testing allocator can't be reset every loop afaik.

Reply:
makes sense - should we make sure `.saftey = true` on the gpa?